EIPS/eip-4788.md

```diff
@@ -77,76 +79,165 @@ When verifying a block, execution clients **MUST** ensure the root value in the
+    set()
+else:
+    get()
 
-timestamp_as_uint256 = to_uint256_be(block_header.timestamp)
-parent_beacon_block_root = block_header.parent_beacon_block_root
+def get():
+    if len(evm.calldata) != 32:
+        evm.revert()
 
-sstore(HISTORY_STORAGE_ADDRESS, timestamp_index, timestamp_as_uint256)
-sstore(HISTORY_STORAGE_ADDRESS, root_index, parent_beacon_block_root)
-```
+    timestamp_idx = to_uint256_be(timestamp) % HISTORICAL_ROOTS_MODULUS
+    timestamp = storage.get(timestamp_idx)
 
-#### New stateful precompile
+    if timestamp != evm.calldata:
+        evm.revert()
 
-Beginning at the execution timestamp `FORK_TIMESTAMP`, a "stateful" precompile is deployed at `HISTORY_STORAGE_ADDRESS`.
+    root_idx = timestamp_idx + HISTORICAL_ROOTS_MODULUS
+    root = storage.get(root_idx)
+    
+    evm.return(root)
 
-Callers of the precompile should provide the `timestamp` they are querying encoded as 32 bytes in big-endian format.
-Clients **MUST** sanitize this input call data to the precompile.
-If the input is _more_ than 32 bytes, the precompile only takes the first 32 bytes of the input buffer and ignores the rest.
-If the input is _less_ than 32 bytes, the precompile should revert.
+def set():
+    timestamp_idx = to_uint256_be(evm.timestamp) % HISTORICAL_ROOTS_MODULUS
+    root_idx = timestamp_idx + HISTORICAL_ROOTS_MODULUS
 
-Given this input, the precompile reduces the `timestamp` in the same way during the write routine and first checks if
-the `timestamp` recorded in the ring buffer matches the one supplied by the caller.
+    storage.set(timestamp_idx, evm.timestamp)
+    storage.set(root_idx, evm.calldata)
+```
 
-If the `timestamp` **does NOT** match, the client **MUST** return the "zero" word -- the 32-byte value where each byte is `0x00`.
+##### Bytecode
+
+The exact initcode to deploy is shared below.
+
+```asm
+push1 0x5a
+dup1
+push1 0x09
+push0
+codecopy
+push0
+return
+
+caller
+push20 0xfffffffffffffffffffffffffffffffffffffffe
+eq
+push1 0x42
+jumpi
+
+push1 0x20
+calldatasize
+eq
+push1 0x24
+jumpi
+
+push0
+push0
+revert
+
+jumpdest
+push3 0x018000
+push0
+calldataload
+mod
+dup1
+sload
+push0
+calldataload
+eq
+iszero
+push1 0x3d
+jumpi
+
+push3 0x018000
+add
+sload
+push0
+mstore
+
+jumpdest
+push1 0x20
+push0
+return
+
+jumpdest
+timestamp
+push3 0x018000
+timestamp
+mod
+sstore
+push0
+calldataload
+push3 0x018000
+timestamp
+mod
+push3 0x018000
+add
+sstore
+stop
+```
 
-If the `timestamp` **does** match, the client **MUST** read the root from the contract storage and return those 32 bytes in the caller's return buffer.
+#### Deployment
+
+The beacon roots contract is deployed like any other smart contract. A special synthetic address is generated
+by working backwards from the desired deployment transaction:
+
+```json
+{
+  "type": "0x2",
+  "chainId": "0x1",
+  "nonce": "0x0",
+  "to": null,
+  "gas": "0xd4f8",
+  "gasPrice": null,
+  "maxPriorityFeePerGas": "0x9c7652400",
+  "maxFeePerGas": "0xe8d4a51000",
```

Comment:
The semi-problem with this fake sig tx is (as raised by @holiman in a discussion we had), that if you want to deploy on other chains, all items are fixed, so also these gas fees, which might be problematic to deploy it on another chain.

Reply:
We could do make uint256 for `maxFeePerGas` if we use an EIP-1559 tx. Ultimately it isn't important how other chains deploy the opcode. The EIP specifies the mainnet behavior, but I assume the address will ultimately become a genesis configuration option much like the deposit contract. Chains could easily etch the code directly into the state by an irregular state transition or via genesis enshrinement.

Reply:
Ok that actually makes a lot of sense!

Reply:
(Small side note: we actually cannot do max uint256 since this violates the balance check of 1559. But this does not matter because you are right, other chains can just "expand" this EIP to just deploy it anywhere as a special state transition)

Reply:
good point 🫠